src/core/packages/saved-objects/api-server-internal/src/lib/apis/search.ts

```diff
@@ -0,0 +1,202 @@
+  const typesToAuthorize = new Set(types);
+  const authorizationResult = await securityExtension?.authorizeFind({
+    namespaces: spacesToAuthorize,
+    types: typesToAuthorize,
+  });
+  if (authorizationResult?.status === 'unauthorized') {
+    // If the user is unauthorized to find *anything* they requested, return an empty response
+    return createEmptySearchResponse();
+  }
+
+  let typeToNamespacesMap: Map<string, string[]> | undefined;
+  if (authorizationResult?.status === 'partially_authorized') {
+    typeToNamespacesMap = new Map<string, string[]>();
+    for (const [objType, entry] of authorizationResult.typeMap) {
+      if (!entry.find) continue;
+      // This ensures that the query DSL can filter only for object types that the user is authorized to access for a given space
+      const { authorizedSpaces, isGloballyAuthorized } = entry.find;
+      typeToNamespacesMap.set(objType, isGloballyAuthorized ? namespaces : authorizedSpaces);
+    }
+  }
+
+  const query = deepMerge(
+    esOptions.query ?? {},
+    getNamespacesBoolFilter({ namespaces, registry, types, typeToNamespacesMap }),
+    {
+      // merge the filter array by putting our namespace filter first
+      arrayMerge: (target, source) => source.concat(target),
+    }
+  );
+
+  const result = await client.search<T, A>(
+    {
+      ...esOptions,
+      // If `pit` is provided, we drop the `index`, otherwise ES returns 400.
+      index: esOptions.pit ? undefined : commonHelper.getIndicesForTypes(types),
+      query,
+    },
+    {
+      ignore: [404],
+      meta: true,
+    }
+  );
+  if (result.statusCode === 404) {
+    if (!isSupportedEsServer(result.headers)) {
+      throw SavedObjectsErrorHelpers.createGenericNotFoundEsUnavailableError();
+    }
+    // 404 is only possible here if the index is missing, which
+    // we don't want to leak, see "404s from missing index" above
+    return createEmptySearchResponse();
+  }
+
+  const redactTypeMapParams: GetFindRedactTypeMapParams = {
+    previouslyCheckedNamespaces: spacesToAuthorize,
+    objects: [],
+  };
+  for (const hit of result.body.hits.hits) {
+    if (serializer.isRawSavedObject(hit as SavedObjectsRawDoc)) {
+      redactTypeMapParams.objects.push({
+        type: hit._source!.type,
+        id: hit._id!,
+        existingNamespaces: hit._source!.namespaces ?? [],
+      });
+    }
+  }
+
+  const redactTypeMap = redactTypeMapParams
+    ? await securityExtension?.getFindRedactTypeMap(redactTypeMapParams)
+    : undefined;
+
+  // Migrations and encryption don't work on raw documents. To process them we have
+  // to serialize raw documents to saved objects and then deserialize them back to
+  // raw documents again.
+  const processHit = async (hit: estypes.SearchHit<T>): Promise<estypes.SearchHit<T>> => {
+    if (!serializer.isRawSavedObject(hit as SavedObjectsRawDoc)) {
+      return hit;
+    }
+
+    const savedObject = serializerHelper.rawToSavedObject(hit as SavedObjectsRawDoc);
```

Comment:
This is a good point, and I think it will reduce complexity for callers if we give them a migrated (up-to-date) and decrypted documents. But their queries and aggregations will still need to be backwards compatible (i.e. work on possibly n-1 model versions).

Conceptually, returning "raw objects" is a bit more work (ironically) but it will be more consistent with the queries and aggs consumers write.

Reply:
Just a bit funny that we will likely go `raw SO -> migrated SO -> raw SO -> (caller) back to SO` 😄... Perhaps we should encourage using `_source` rather than converting back to a "normal" SO.